src/Iterators/LDAPIterator.php

```diff
@@ -0,0 +1,294 @@
+
+        ldap_control_paged_result($resource, $this->getPageSize(), true, $this->cookie);
+        if ($this->getReturnAttributes() !== null) {
+            $resultResource = ldap_search($resource, $baseDn, $this->getFilter(), $this->getReturnAttributes());
+        } else {
+            $resultResource = ldap_search($resource, $baseDn, $this->getFilter());
+        }
+        if (! is_resource($resultResource)) {
+            /*
+             * @TODO better exception msg
+             */
+            throw new \Exception('ldap_search returned something wrong...' . ldap_error($resource));
+        }
+
+        $entries = ldap_get_entries($resource, $resultResource);
+        if ($entries === false) {
+            throw new LdapException($ldap, 'Entries could not get fetched');
+        }
+        $entries = $this->getConvertedEntries($entries);
+
+        ErrorHandler::start();
+        $response = ldap_control_paged_result_response($resource, $resultResource, $this->cookie);
+        ErrorHandler::stop();
+
+        if ($response !== true) {
+            throw new LdapException($ldap, 'Paged result was empty');
+        }
+
+        if ($this->entries === null) {
+            $this->entries = [];
+        }
+
+        $this->entries = array_merge($this->entries, $entries);
+
+        return true;
+    }
+    private function getConvertedEntries(array $entries)
+    {
+        $result = [];
+
+        foreach ($entries as $key => $entry) {
+            if ($key === 'count') {
+                continue;
+            }
+
+            $result[$key] = $this->getConvertedEntry($entry);
+        }
+
+        return $result;
+    }
+    private function getConvertedEntry(array $entry)
+    {
+        $result = [];
+
+        foreach ($entry as $key => $value) {
+            if (is_int($key)) {
+                continue;
+            }
+            if ($key === 'count') {
+                continue;
+            }
+
+            if (isset($value['count'])) {
+                unset($value['count']);
+            }
+
+            $result[$key] = $value;
+        }
+
+        if ($this->getResolveRangedAttributes() === true) {
+            $result = $this->resolveRangedAttributes($result);
+        }
+
+        return $result;
+    }
+    private function resolveRangedAttributes(array $row)
+    {
+        $result = [];
+        foreach ($row as $key => $value) {
+            $keyExploded = explode(';range=', $key);
+
+            if (count($keyExploded) === 2) {
+                $range = explode('-', $keyExploded[1]);
+                $offsetAndLimit = (int) $range[1] + 1;
+
+                $result[$keyExploded[0]] = array_merge($value, $this->getAttributeRecursive($row['dn'], $keyExploded[0], $offsetAndLimit, $offsetAndLimit));
+            } else {
+                $result[$key] = $value;
+            }
+        }
+
+        return $result;
+    }
+    private function getAttributeRecursive(string $dn, string $attrName, int $offset, int $maxPerRequest)
```

Comment:
I'm not sure this module is PHP 7 minimum yet so we can't use scalar type hints

Reply:
Sweet, have removed all these

Reply:
I've unresolved a bunch of comments because they don't look to have been done, or maybe got reverted at some point =)